components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt

```diff
@@ -107,19 +111,24 @@ internal class FilePicker(
      * @param intent The result of the request.
      */
     fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
-        if (requestCode == FILE_PICKER_ACTIVITY_REQUEST_CODE) {
+        val request = getActivePromptRequest() ?: return
+        if (requestCode == FILE_PICKER_ACTIVITY_REQUEST_CODE && request is File) {
             store.consumePromptFrom(sessionId) {
```

Comment:
So looking at this flow, what could happen here (and be the cause of the problem) is that the selected session changed and we're now processing a prompt request from a different session. It's one possibility. I think, ideally, we should make this impossible by creating a `FilePicker` instance for the specific session/prompt request (and store the sessionId of the selected session or custom tab), instead of using a "global" file picker instance which looks up the prompt request for the selected session which could have changed...

Reply:
Yes that could be the issue, we request the OS to show the file picker but it took a lot of time, the user select a new session, and this session has another dialog, and then the file picker appears, the user select the file and we end up trying to consume a different prompt request.

It's a bit tricky to store the `sessionId` in the in the file picker dialog, because we don't own it, we request the OS to provide us an app that can handle the file picker functionality ([here](https://github.com/mozilla-mobile/android-components/blob/0c856f9584cadd3d57081b5c0304595a4974291c/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt#L74) and [here](https://github.com/mozilla-mobile/android-components/blob/0c856f9584cadd3d57081b5c0304595a4974291c/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt#L100)), via an intent, but there is not an additional information that we can use to identify the session when the OS replies back to `onActivityResult`, some information that is replied back from [startActivityForResult](https://developer.android.com/reference/android/app/Activity#startActivityForResult(android.content.Intent,%20int)) to [onActivityResult](https://developer.android.com/reference/android/app/Activity#onActivityResult(int,%20int,%20android.content.Intent)) is the `requestCode` we could use the `sessionId` instead of the [FILE_PICKER_ACTIVITY_REQUEST_CODE](https://github.com/mozilla-mobile/android-components/blob/0c856f9584cadd3d57081b5c0304595a4974291c/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/file/FilePicker.kt#L100).

Reply:
@Amejia481 

> It's a bit tricky to store the sessionId in the in the file picker dialog

Ah no, I meant storing it as a field in `FilePicker` similar to how it stores the `customTabSessionId`: https://github.com/mozilla-mobile/android-components/blob/master/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt#L218.

Reply:
Got it, I will take a look!